lib/elixir/lib/kernel.ex

```diff
@@ -2198,6 +2198,54 @@ defmodule Kernel do
     :erlang.error(ArgumentError.exception(error_message))
   end
 
+  @doc """
+  Returns true if `term` is a struct; otherwise returns `false`.
+
+  Allowed in guard tests.
+  """
+  @doc guard: true
+  defmacro is_struct(term) do
+    case __CALLER__.context do
+      nil ->
+        quote do
+          case unquote(term) do
+            %_{} -> true
+            _ -> false
+          end
+        end
+
+      :match ->
+        invalid_match!(:is_struct)
+
+      :guard ->
+        quote(do: is_atom(:erlang.map_get(:__struct__, unquote(term))))
+    end
+  end
+
+  @doc """
+  Returns true if `term` is a struct; otherwise returns `false`.
+
+  Allowed in guard tests.
+  """
+  @doc guard: true
+  defmacro is_struct(term, struct) do
```

Comment:
I think we should rename the second argument to `struct_type` or `struct_name`, because it is confusion, and it could be interpreted as an actual struct is needed to be passed. Actually, `Kernel.defstruct/2` works this very same way, it accepts its second argument `struct` either an atom or a struct.